src/OSPSuite.Presentation/Presenters/Charts/ChartEditorPresenter.cs

```diff
@@ -240,15 +247,80 @@ public class ChartEditorPresenter : AbstractCommandCollectorPresenter<IChartEdit
 
          _view.SetAxisSettingsView(axisSettingsPresenter.View);
          _view.SetChartSettingsView(chartSettingsPresenter.View);
          _view.SetChartExportSettingsView(chartExportSettingsPresenter.View);
          _view.SetCurveSettingsView(curveSettingsPresenter.View);
+         _view.SetCurveColorGroupingView(curveColorGroupingPresenter.View);
          _view.SetDataBrowserView(dataBrowserPresenter.View);
       }
 
+      //gets all the common metaData of the observed data that correspond to active curves 
+      private IReadOnlyList<string> getCommonMetaDataOfCurves(IReadOnlyList<DataRepository> allObservedData)
+      {
+         var activeDataColumns = _dataBrowserPresenter.GetAllUsedDataColumns();
+         _activeObservedData = allObservedData.Where(x => activeDataColumns.Any(column => column.Repository.Name == x.Name)).ToList();
+
+         if (_activeObservedData == null || !_activeObservedData.Any())
+            return new List<string>();
+
+         var firstCurveMetaData = _activeObservedData.First().ExtendedProperties.Keys.ToList();
+         var commonMetaData = firstCurveMetaData.Where(x => _activeObservedData.All(observedData => observedData.ExtendedProperties.Keys.Contains(x))).ToList();
+
+         return commonMetaData;
+      }
+
+      private void onApplyColorGrouping(IReadOnlyList<string> eSelectedMetaData)
+      {
+         if (_activeObservedData == null)
+            return;
+
+         var groupedDataRepositories = GroupDataRepositories(eSelectedMetaData);
+         assignSameColorToGroupedCurves(groupedDataRepositories);
+      }
+
+      private void assignSameColorToGroupedCurves(List<List<DataRepository>> groupedDataRepositories)
+      {
+         foreach (var group in groupedDataRepositories)
+         {
+            var groupColor = Chart.SelectNewColor();
+            var curvesInGroup = findCurvesCorrespondingToRepositoryList(group);
+
+            if (!curvesInGroup.Any())
+               continue;
+            
+            curvesInGroup.Each(x => _curveSettingsPresenter.UpdateColorForCurve(x, groupColor));
+         }
+      }
+
+      private List<List<DataRepository>> GroupDataRepositories(IReadOnlyList<string> groupingCriteria)
+      {
+         var groupedDataRepositories = new List<List<DataRepository>> { _activeObservedData };
+         foreach (var groupingMetaData in groupingCriteria)
+         {
+            var tempGroupedList = new List<List<DataRepository>>();
+            foreach (var existingGroup in groupedDataRepositories)
+            {
+               var dataReposGroupedBySingleMetaData =
+                  existingGroup.GroupBy(x => x.ExtendedProperties[groupingMetaData].ValueAsObject).Select(group => @group.ToList());
+               tempGroupedList.AddRange(dataReposGroupedBySingleMetaData);
+            }
+
+            groupedDataRepositories = tempGroupedList;
```

Comment:
I don't quite understand this. Aren't you overriding this list EVERY time you itereate int eh first for Loop?

Reply:
Yeah I think this is worth some discussing. This is was my best idea algorithmically to take advantage of .GroupBy() to group by multiple criteria:

I group by the first criterion and I get a IEnumerable<List<DataRepository>> containing the groups. If I were to apply .GroupBy() again to each group I would get IEnumerable<IEnumerable<List<DataRepository>>> and so on. So at the end of each grouping I need to "flatten" the structure, since we only care that they are separate groups, we do not really care what the "parent" group is. To do that I am using this temp list.

Reply:
lol I still don't get it. And I am trying. I think I need to see it in action. I am going to get the code and run it.
How can I get the code to activate in the test program?